### PR TITLE
Add upgrade check rule for unrecognized arguments to Operators

### LINKED ISSUE #11042: Create NoAdditionalArgsInOperatorsRule to ease upgrade to Airflow 2.0
This issue is part of #8765

## Rule
Create `NoAdditionalArgsInOperatorsRule` which corresponds to
> Additional arguments passed to BaseOperator cause an exception

entry in UPDATING.md. This rule should allow users to check if their current configuration needs any adjusting
before migration to Airflow 2.0.

## How to guide
To implement a new rule, create a class that inherits from `airflow.upgrade.rules.base_rule.BaseRule`.
It will be auto-registered and used by `airflow upgrade-check` command. The custom rule class has to have `title`,
`description` properties and should implement `check` method which returns a list of error messages in case of
incompatibility.

For example:
https://github.com/apache/airflow/blob/ea36166961ca35fc51ddc262ec245590c3e236fb/airflow/upgrade/rules/conn_type_is_not_nullable.py#L25-L42

**Remember** to open the PR against `v1-10-test` branch.

### PR DESCRIPTION
**Pull request by drdenzy**
ease upgrade to Airflow 2.0

The NoAdditionalArgsInOperatorsRule class ensures that passing an
unrecognised arguments to operators causes an exception.

Passing unrecognised arguments is not allowed in Airflow 2.0

fixes #11042

### CODE REVIEW

File: airflow/upgrade/rules/no_additional_args_in_operators.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+from airflow.upgrade.rules.base_rule import BaseRule
+
+
+class NoAdditionalArgsInOperatorsRule(BaseRule):
+    title = "No additional argument allowed in BaseOperator."
+
+    description = """
+                Passing unrecognized arguments to operators is not allowed in Airflow 2.0 anymore,
+                and will cause an exception.
+                  """
```
Comment (turbaszek) [inline airflow/upgrade/rules/no_additional_args_in_operators.py]:
```suggestion
    description = """\
Passing unrecognised arguments to operators is not allowed in Airflow 2.0 anymore,
and will cause an exception.
"""
```

To avoid leading new line and indentation

File: airflow/upgrade/rules/no_additional_args_in_operators.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    def check(self, dags_folder=None):
+        if not dags_folder:
+            dags_folder = conf.get("core", "dags_folder")
```
Comment (turbaszek) [inline airflow/upgrade/rules/no_additional_args_in_operators.py:40]:
```suggestion
    def check(self):
        dags_folder = conf.get("core", "dags_folder")
```
I don't think users will be able to specify the dag folder in other way

File: airflow/upgrade/rules/no_additional_args_in_operators.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+        logger = logging.root
+        old_level = logger.level
+        try:
+            logger.setLevel(logging.ERROR)
```
Comment (turbaszek) [inline airflow/upgrade/rules/no_additional_args_in_operators.py:45]: If this is here to suppress the log then it won't be necessary after https://github.com/apache/airflow/pull/12657

File: airflow/upgrade/rules/no_additional_args_in_operators.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                try:
+                    with warnings.catch_warnings(record=True) as captured_warnings:
+                        _ = dagbag.process_file(
+                            filepath,
+                            safe_mode=conf.getboolean('core', 'DAG_DISCOVERY_SAFE_MODE'))
+                except Exception:
+                    pass
+
+                for warning in captured_warnings:
+                    if warning.category == PendingDeprecationWarning \
```
Comment (turbaszek) [inline airflow/upgrade/rules/no_additional_args_in_operators.py]:
```suggestion
                    if warning.category in (DeprecationWarning, PendingDeprecationWarning) \
```

File: airflow/upgrade/rules/no_additional_args_in_operators.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                            and str(warning.message).startswith("Invalid arguments were passed"):
+                        m = re.match(r'''
+                                    .* \(task_id:\ ([^\)]+)\) .* \n
+                                    \*args:\ (.*) \n
+                                     \*\*kwargs:\ (.*)
+                                    ''', str(warning.message), re.VERBOSE)
```
Comment (turbaszek) [inline airflow/upgrade/rules/no_additional_args_in_operators.py:68]: I'm not sure about the indentation in this regexp (due to multiline string)

File: tests/upgrade/rules/test_no_additional_args_operators.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+from tempfile import NamedTemporaryFile
+from unittest import TestCase
+import textwrap
+from airflow.upgrade.rules.no_additional_args_in_operators import NoAdditionalArgsInOperatorsRule
+from tests.test_utils.db import clear_db_connections
+
+
+class TestNoAdditionalArgsInOperatorsRule(TestCase):
+    def tearDown(self):
+        clear_db_connections()
```
Comment (turbaszek) [inline tests/upgrade/rules/test_no_additional_args_operators.py]: Is it necessary?

File: tests/upgrade/rules/test_no_additional_args_operators.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            from airflow.utils.dates import days_ago
+            from airflow.operators.bash_operator import BashOperator
+
+            with DAG(dag_id="test", start_date=days_ago(0)):
+                BashOperator(task_id='test', bash_command="true", extra_param=42)
+                '''
+                                ))
+            dag_file.flush()
+            msgs = rule.check(dags_folder=dag_file.name)
+            assert list(msgs) != []
```
Comment (turbaszek) [inline tests/upgrade/rules/test_no_additional_args_operators.py]:
```suggestion
            assert len(list(msgs)) == 1
```
Right?

File: tests/upgrade/rules/test_no_additional_args_operators.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            from airflow import DAG
+            from airflow.utils.dates import days_ago
+            from airflow.operators.bash_operator import BashOperator
+
+            with DAG(dag_id="test", start_date=days_ago(0)):
+                BashOperator(task_id='test', bash_command="true", extra_param=42)
+                '''
+                                ))
+            dag_file.flush()
+            msgs = rule.check(dags_folder=dag_file.name)
```
Comment (turbaszek) [inline tests/upgrade/rules/test_no_additional_args_operators.py]:
```suggestion
            msgs = rule.check()
```
If we have this argument only for test then we should remove it and use mock similar to this:
https://github.com/apache/airflow/blob/847820f91207dcee815ecda96e48cf62b0977580/tests/upgrade/rules/test_import_changes.py#L49-L61

File: airflow/upgrade/rules/no_additional_args_in_operators.py (hunk 1 above)
Comment (drdenzy) [reply airflow/upgrade/rules/no_additional_args_in_operators.py]: Oh yeah! Thanks for pointing that out.

File: airflow/upgrade/rules/no_additional_args_in_operators.py (hunk 2 above)
Comment (drdenzy) [reply airflow/upgrade/rules/no_additional_args_in_operators.py:40]: ok.

File: airflow/upgrade/rules/no_additional_args_in_operators.py (hunk 4 above)
Comment (drdenzy) [reply airflow/upgrade/rules/no_additional_args_in_operators.py]: I agree 👍

File: airflow/upgrade/rules/no_additional_args_in_operators.py (hunk 5 above)
Comment (drdenzy) [reply airflow/upgrade/rules/no_additional_args_in_operators.py:68]: That's where the `re.VERBOSE` comes in I believe.

File: tests/upgrade/rules/test_no_additional_args_operators.py (hunk 6 above)
Comment (drdenzy) [reply tests/upgrade/rules/test_no_additional_args_operators.py]: Thanks for pointing that out. I should have taken that out after I made later changes.

File: tests/upgrade/rules/test_no_additional_args_operators.py (hunk 7 above)
Comment (drdenzy) [reply tests/upgrade/rules/test_no_additional_args_operators.py]: Yeah 👍

File: tests/upgrade/rules/test_no_additional_args_operators.py (hunk 8 above)
Comment (drdenzy) [reply tests/upgrade/rules/test_no_additional_args_operators.py]: Noted.

File: airflow/upgrade/rules/no_additional_args_in_operators.py (hunk 2 above)
Comment (ashb) [reply airflow/upgrade/rules/no_additional_args_in_operators.py:40]: This is needed for testing, so leave it for now @Dr-Denzy

Comment (ashb) [reply airflow/upgrade/rules/no_additional_args_in_operators.py:40]: This is a similar approach that was taken in the https://github.com/apache/airflow/blob/f47efe91a8b4ba5ba43fa8d55990d27516ab0b6c/airflow/upgrade/rules/undefined_jinja_varaibles.py#L133

File: airflow/upgrade/rules/no_additional_args_in_operators.py (hunk 5 above)
Comment (ashb) [reply airflow/upgrade/rules/no_additional_args_in_operators.py:68]: Yeah -- re.VERBOSE means whitespace in the pattern is ignored -- we can indent it how we like, or even add comments.

File: tests/upgrade/rules/test_no_additional_args_operators.py (hunk 8 above)
Comment (ashb) [reply tests/upgrade/rules/test_no_additional_args_operators.py]: Ah, the Dependency Injection vs Mocking debate. :)

File: airflow/contrib/kubernetes/pod.py [hunk 9]
```diff
@@ -250,8 +250,8 @@ def _extract_ports(ports):
 
 def _extract_resources(resources):
     if isinstance(resources, k8s.V1ResourceRequirements):
-        requests = resources.requests
-        limits = resources.limits
+        requests = resources.requests or {}
+        limits = resources.limits or {}
```
Comment (ashb) [inline airflow/contrib/kubernetes/pod.py]: You somehow got some extra unrelated changes committted.

Comment (kaxil) [reply airflow/contrib/kubernetes/pod.py]: fixed

File: tests/upgrade/rules/test_no_additional_args_operators.py (hunk 8 above)
Comment (ashb) [reply tests/upgrade/rules/test_no_additional_args_operators.py]: Having slept on it, I personally find the DI approach easier to understand than mocking.

Comment (turbaszek) [reply tests/upgrade/rules/test_no_additional_args_operators.py]: @ashb personally I don't like features in code that are not used when running the code - it creates confusion when not commented

Comment (ashb) [reply tests/upgrade/rules/test_no_additional_args_operators.py]: Also true.

Comment (ashb) [reply tests/upgrade/rules/test_no_additional_args_operators.py]: @turbaszek Since you and I don't agree, and both approaches already exist in the code base, are you happy to let @Dr-Denzy decide on which approach to user?

File: tests/upgrade/rules/test_no_additional_args_operators.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+            from airflow import DAG
+            from airflow.utils.dates import days_ago
+            from airflow.operators.bash_operator import BashOperator
+
+            with DAG(dag_id="test", start_date=days_ago(0)):
+                BashOperator(task_id='test', bash_command="true", extra_param=42)
+                '''))
+            dag_file.flush()
+            msgs = rule.check(dags_folder=dag_file.name)
+            assert len(list(msgs)) == 1
```
Comment (ashb) [inline tests/upgrade/rules/test_no_additional_args_operators.py]:
```suggestion
            msgs = list(rule.check(dags_folder=dag_file.name))
            assert len(msgs) == 1
```

This might give us some indication what the failure on CI is

**Comment by kaxil** [comment]
Weird, passed for me locally too

**Comment by potiuk** [comment]
> Weird, passed for me locally too

As discussed here: https://apache-airflow.slack.com/archives/C0146STM600/p1606855518349800 - I belive this might side-effect from another test using the same "test" dag_id. We have a few of those and there are a number of tests that are likely leaving some remnants in the DB. My guess would be, this is the case and cleaning in setUp/tearDown might solve the issue.

File: airflow/upgrade/rules/no_additional_args_in_operators.py (hunk 3 above)
Comment (ashb) [reply airflow/upgrade/rules/no_additional_args_in_operators.py:45]: @Dr-Denzy We can remove the logger.setLevel, and the outer try/finally now, as the PR mentioned here has been merged

File: tests/upgrade/rules/test_no_additional_args_operators.py (hunk 8 above)
Comment (turbaszek) [reply tests/upgrade/rules/test_no_additional_args_operators.py]: > @turbaszek Since you and I don't agree, and both approaches already exist in the code base, are you happy to let @Dr-Denzy decide on which approach to user?

Yes, no objection to keep it as it is - we already use this approach 👌

**Comment by turbaszek** [comment]
> > Weird, passed for me locally too
> 
> As discussed here: https://apache-airflow.slack.com/archives/C0146STM600/p1606855518349800 - I belive this might side-effect from another test using the same "test" dag_id. We have a few of those and there are a number of tests that are likely leaving some remnants in the DB. My guess would be, this is the case and cleaning in setUp/tearDown might solve the issue.

Sounds familiar. I usually prefer to use tests names as dag/task ids as this make sure that we are not affected by side effects. But cleanup would be also nice.

**Review by kaxil — APPROVED** [review]

File: tests/upgrade/rules/test_no_additional_args_operators.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+from unittest import TestCase
+
+import pytest
+
+from airflow.upgrade.rules.no_additional_args_in_operators import NoAdditionalArgsInOperatorsRule
+
+
+class TestNoAdditionalArgsInOperatorsRule(TestCase):
+
+    @pytest.mark.filterwarnings('always')
```
Comment (kaxil) [inline tests/upgrade/rules/test_no_additional_args_operators.py:30]: This was needed as we suppress DeprecationWarning and PendingDeprecationWarning when running pytest on CI:

```
Starting the tests with those pytest arguments: --verbosity=0 --strict-markers --durations=100 --cov=airflow/ --cov-config=.coveragerc --cov-report=xml:/files/coverage.xml --color=yes --maxfail=50 --pythonwarnings=ignore::DeprecationWarning --pythonwarnings=ignore::PendingDeprecationWarning --junitxml=/files/test_result.xml --setup-timeout=20 --execution-timeout=60 --teardown-timeout=20 -rfEX --with-db-init tests
```

Lol took few hours for me & Dennis to figure why we were not able to replicate it when running it with breeze:

```
root@e6768532ef19:/opt/airflow# pytest tests/upgrade/rules/test_no_additional_args_operators.py
```

until we ran it with:

```
root@e6768532ef19:/opt/airflow# pytest tests/upgrade/rules/test_no_additional_args_operators.py --pythonwarnings=ignore::DeprecationWarning --pythonwarnings=ignore::PendingDeprecationWarning
```

😄

Comment (ashb) [reply tests/upgrade/rules/test_no_additional_args_operators.py:30]: Good work you two!
